Userland/Libraries/LibAudio/PlaybackStreamOboe.cpp

```diff
@@ -0,0 +1,151 @@
+    {
+    }
+    Duration last_sample_time() const
+    {
+        return Duration::from_seconds(m_last_sample_time.load());
+    }
+    void set_volume(float volume)
+    {
+        m_volume.store(volume);
+    }
+
+private:
+    PlaybackStream::AudioDataRequestCallback m_data_request_callback;
+    Atomic<i64> m_last_sample_time { 0 };
+    size_t m_number_of_samples_enqueued { 0 };
+    Atomic<float> m_volume { 1.0 };
+};
+
+class PlaybackStreamOboe::Storage : public RefCounted<PlaybackStreamOboe::Storage> {
+public:
+    Storage(std::shared_ptr<oboe::AudioStream> stream, std::shared_ptr<OboeCallback> oboe_callback)
+        : m_stream(move(stream))
+        , m_oboe_callback(move(oboe_callback))
+    {
+    }
+    std::shared_ptr<oboe::AudioStream> stream() const { return m_stream; }
+    std::shared_ptr<OboeCallback> oboe_callback() const { return m_oboe_callback; }
+
+private:
+    std::shared_ptr<oboe::AudioStream> m_stream;
+    std::shared_ptr<OboeCallback> m_oboe_callback;
+};
+
+PlaybackStreamOboe::PlaybackStreamOboe(NonnullRefPtr<Storage> storage)
+    : m_storage(move(storage))
+{
+}
+
+ErrorOr<NonnullRefPtr<PlaybackStream>> PlaybackStreamOboe::create(OutputState initial_output_state, u32 sample_rate, u8 channels, u32, AudioDataRequestCallback&& data_request_callback)
+{
+    std::shared_ptr<oboe::AudioStream> stream;
+    auto oboe_callback = std::make_shared<OboeCallback>(move(data_request_callback));
+    oboe::AudioStreamBuilder builder;
+    auto result = builder.setSharingMode(oboe::SharingMode::Shared)
+                      ->setPerformanceMode(oboe::PerformanceMode::LowLatency)
+                      ->setFormat(oboe::AudioFormat::Float)
+                      ->setDataCallback(oboe_callback)
+                      ->setChannelCount(channels)
+                      ->setSampleRate(sample_rate)
+                      ->openStream(stream);
+
+    if (result != oboe::Result::OK)
+        return Error::from_string_literal("Oboe failed to start");
+
+    if (initial_output_state == OutputState::Playing)
+        stream->requestStart();
+
+    auto storage = TRY(adopt_nonnull_ref_or_enomem(new PlaybackStreamOboe::Storage(move(stream), move(oboe_callback))));
+    return TRY(adopt_nonnull_ref_or_enomem(new (nothrow) PlaybackStreamOboe(move(storage))));
+}
+
+PlaybackStreamOboe::~PlaybackStreamOboe() = default;
+
+void PlaybackStreamOboe::set_underrun_callback(Function<void()>)
+{
+    // FIXME: Implement this.
+}
+
+NonnullRefPtr<Core::ThreadedPromise<Duration>> PlaybackStreamOboe::resume()
+{
+    auto promise = Core::ThreadedPromise<Duration>::create();
+    auto time = MUST(total_time_played());
+    m_storage->stream()->requestStart();
+    promise->resolve(move(time));
+    return promise;
+}
+
+NonnullRefPtr<Core::ThreadedPromise<void>> PlaybackStreamOboe::drain_buffer_and_suspend()
+{
+    auto promise = Core::ThreadedPromise<void>::create();
+    m_storage->stream()->requestPause();
```

Comment:
Does the `requestPause()` function stop the callback being called and play all samples that have buffered before completing the task? Without doing so, we may have some leftover audio in the buffers that will play upon resuming.

We should also be in that paused state before the promise is resolved.

Reply:
Fixed, now it uses stop, which plays the entire buffer